### PR TITLE
add Tree-Path KL Divergence loss for hier classification + unit test

### PR DESCRIPTION
**Pull request by Jyc323**
### Summary
Introduced a new hierarchical regularization loss `TreePathKLDivergenceLoss` that enforces consistency across hierarchical label predictions by minimizing KL divergence between model probabilities and a constructed target distribution along the true label path.

### Key Changes
- Added `tree_path_kl_divergence.py` under `lib/src/otx/backend/native/models/common/losses/`
  - Implements `TreePathKLDivergenceLoss` with configurable reduction (`batchmean`, `mean`, `sum`) and weighting.
  - Target distribution assigns equal probability (1/L) to the GT class at each hierarchy level.
- Added comprehensive unit tests in `lib/tests/unit/backend/native/models/common/losses/test_tree_path_kl_divergence.py` covering:
  - Shape and finiteness checks across multiple levels.
  - Gradient flow validation to ensure backprop works correctly.
  - Alignment vs misalignment sanity check: aligned logits give lower loss.
  - **Exact-value correctness** for:
    - Single-level case (loss equals `CrossEntropyLoss`).
    - Multi-level case (manual KL computation matches PyTorch loss).

### Motivation
Existing metrics capture evaluation aspects of hierarchical classification, but training lacked a regularizer that explicitly
encourages consistency across levels. The Tree-Path KL loss:
- Penalizes predictions that scatter probability mass off the ground-truth path.
- Provides a smooth differentiable signal to improve hierarchical consistency beyond hard CE losses.

This makes the training loop more robust for hierarchical models, reducing path-inconsistent predictions and improving downstream metrics.

### Preliminary Results
We evaluated the effect of Tree-Path KL loss on multiple hierarchical datasets using **EfficientNet_v2 (timm model)** from OTX. The KL weight (λ) was varied between 0 (baseline), 1, and 5. Results show that incorporating KL improves performance on most datasets.

| Dataset          | Metric             | λ=0  | λ=1  | λ=5  |
|------------------|--------------------|-------|-------|-------|
| **CUB_200_2011** | Accuracy           | 0.853 | 0.850 | **0.858** |
|                  | Full Path Accuracy | 0.735 | 0.724 | **0.764** |
|                  | Leaf Accuracy      | **0.825** | 0.821 | 0.812 |
|                  | Weighted Precision | **0.836** | 0.832 | 0.829 |
| **AmazonParrot** | Accuracy           | 0.718 | 0.725 | **0.742** |
|                  | Full Path Accuracy | 0.611 | 0.614 | **0.659** |
|                  | Leaf Accuracy      | 0.333 | 0.334 | **0.346** |
|                  | Weighted Precision | 0.490 | **0.504** | **0.504** |
| **FGVC-Aircraft**| Accuracy           | 0.709 | 0.742 | **0.764** |
|                  | Full Path Accuracy | 0.551 | 0.593 | **0.629** |
|                  | Leaf Accuracy      | 0.614 | 0.639 | **0.660** |
|                  | Weighted Precision | 0.705 | 0.730 | **0.746** |
| **Fish-Vista**   | Accuracy           | 0.460 | 0.497 | **0.507** |
|                  | Full Path Accuracy | 0.217 | 0.264 | **0.280** |
|                  | Leaf Accuracy      | 0.416 | 0.452 | **0.458** |
|                  | Weighted Precision | 0.399 | 0.435 | **0.438** |

**Observations:**
- Increasing KL weight to 5 often improves **Full Path Accuracy**, showing stronger hierarchical consistency.
- Gains are consistent across datasets of varying granularity (birds, parrots, aircraft, fish).
- Some trade-off exists in **Leaf Accuracy** (CUB shows slight drop), but overall performance trends upward in hierarchy-sensitive metrics.

These results indicate that Tree-Path KL is a promising regularizer for hierarchical classification tasks.

### How to test

```
 pytest -q lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py
```

### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [x] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/open-edge-platform/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/open-edge-platform/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/open-edge-platform/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by rajeshgangireddy** [comment]
PS : This and  #4689  are PRs from our GSOC contributor.
Part of the project was 'H-label classification for VLMs' but the metrics and loss functions are useful and would be a nice addition to OTX.

File: lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    log_p[mask] = torch.log(p[mask])
+    manual_kl = (p * (log_p - log_q)).sum() / B
+
+    # Loss under test (must match manual)
+    loss_fn = TreePathKLDivergenceLoss(reduction="batchmean")
+    test_kl = loss_fn([logits0.float(), logits1.float()], targets)
+
+    assert torch.allclose(test_kl.double(), manual_kl, atol=1e-8), (
+        f"manual={manual_kl.item():.12f} vs loss={test_kl.item():.12f}"
+    )
```
Comment (sovrasov) [inline lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py]: Missing new line at EOF

File: lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""Module for defining TreePathKLDivergenceLoss."""
+
+from __future__ import annotations
+from typing import List
```
Comment (sovrasov) [inline lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py]: We're using python > 3.10, there is no need to import `List`: built-in `list` can be used instead

File: lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2022 Intel Corporation
```
Comment (sovrasov) [inline lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py]: 2025

File: lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
+import torch
```
Comment (sovrasov) [inline lib/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py]: License header is missing

File: lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    Inputs:
+        logits_list: list of tensors [B, C_l], ordered from root -> leaf
+        targets: LongTensor [B, L] with per-level GT indices (L == len(logits_list))
+
+    The target distribution places 1/L probability on the GT index for each level,
+    and 0 elsewhere, then uses KLDivLoss(log_softmax(logits), target_probs).
+    """
+
+    def __init__(self, reduction: str = "batchmean", loss_weight: float = 1.0):
```
Comment (sovrasov) [inline lib/src/otx/backend/native/models/classification/losses/tree_path_KL_divergence_loss.py]: If the intention is to follow KL loss API, `reduction` should be `str | None = "batchmean"`

**Review by sovrasov — CHANGES_REQUESTED** [review]
Thanks @Jyc323 for you contribution, I have a question:
- This PR adds a new loss implementation but does not reveal how OTX users can utilize that loss when solving their tasks. Could you also add an integration into existing h-cls training pipeline and mention that improvement in OTX documentation? I believe loss choice should be configurable via OTX model recipe.

**Comment by sovrasov** [comment]
Code quality checks can be launched locally with `cd lib && tox r -vv -e pre-commit`

**Comment by Jyc323** [comment]
Hi @kprokofi, as suggested, I created a new example recipe yaml file `efficientnet_v2_kl.yaml` for integrating tree path KL divergence loss. To successfully running the recipe file with engine, I added a new model class `TimmModelHLabelClsWithKL` under `lib/src/otx/backend/native/models/classification/hlabel_models/timm_model.py`, and a new classifier class `KLHLabelClassifier` under `lib/src/otx/backend/native/models/classification/classifier/h_label_classifier.py`. Could you please take a look and share any suggestions?

While running `tox r -vv -e pre-commit`, I encountered an error:
> lib/src/otx/backend/native/utils/utils.py:89: error: Module has no attribute "metrics"  [attr-defined]
> Found 1 error in 1 file (checked 363 source files)git

I'm not sure if it's caused by something on my end. 

Also, could you let me know the appropriate location for the documentation? If possible, please provide a template or example I can refer to.

Thanks!

**Comment by Jyc323** [comment]
Hi @sovrasov, I merged this PR with the current develop branch. Please let me know the next step. Thanks!

File: library/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
+# Copyright (C) 2025 Intel Corporation
```
Comment (sovrasov) [inline library/tests/unit/backend/native/models/classification/losses/test_tree_path_kl_divergence.py:1]: This file should be moved to `library/tests/...`, since `lib` folder was renamed to `library`

File: library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py [hunk 7]
```diff
@@ -89,3 +89,66 @@ def _create_model(self, head_config: dict | None = None) -> nn.Module:  # type:
     def load_from_otx_v1_ckpt(self, state_dict: dict, add_prefix: str = "model.") -> dict:
         """Load the previous OTX ckpt according to OTX2.0."""
         return OTXv1Helper.load_cls_effnet_v2_ckpt(state_dict, "hlabel", add_prefix)
+
+
+class TimmModelHLabelClsWithKL(OTXHlabelClsModel):
```
Comment (sovrasov) [inline library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py]: It looks like a similar classes should be implemented for all the model files in this folder.

I was quickly thinking about how can we avoid that, but my only thought for now is a refactoring of the parent h-cls classes introducing classifier type generics to generate a bunch of h-cls model classes with different h-label classifiers. That's a solution from the top of my head, one can settle on simple duplication for now.

File: library/src/otx/backend/native/models/classification/classifier/h_label_classifier.py [hunk 8]
```diff
@@ -143,3 +144,87 @@ def _forward_explain(self, images: torch.Tensor) -> dict[str, torch.Tensor | lis
             outputs["preds"] = preds
 
         return outputs
+
+
+class KLHLabelClassifier(HLabelClassifier):
```
Comment (sovrasov) [inline library/src/otx/backend/native/models/classification/classifier/h_label_classifier.py:149]: New classifier stuff should also be covered with UTs

**Comment by sovrasov** [comment]
> Hi @kprokofi, as suggested, I created a new example recipe yaml file `efficientnet_v2_kl.yaml` for integrating tree path KL divergence loss. To successfully running the recipe file with engine, I added a new model class `TimmModelHLabelClsWithKL` under `lib/src/otx/backend/native/models/classification/hlabel_models/timm_model.py`, and a new classifier class `KLHLabelClassifier` under `lib/src/otx/backend/native/models/classification/classifier/h_label_classifier.py`. Could you please take a look and share any suggestions?
> 
> While running `tox r -vv -e pre-commit`, I encountered an error:
> 
> > lib/src/otx/backend/native/utils/utils.py:89: error: Module has no attribute "metrics"  [attr-defined]
> > Found 1 error in 1 file (checked 363 source files)git
> 
> I'm not sure if it's caused by something on my end.
> 
> Also, could you let me know the appropriate location for the documentation? If possible, please provide a template or example I can refer to.
> 
> Thanks!

Thanks @Jyc323 for the recent updates. Once new classifier is available for each classification models and extra unit tests are added, we're almost there: only docs are to be added.
Since we have [an advanced tutorial](https://github.com/open-edge-platform/training_extensions/blob/develop/library/docs/source/guide/tutorials/advanced/hier_metric_collection.rst) for h-cls metrics, we can extend it to an advanced h-cls tutorial by introducing examples on how to use the new loss with the new metrics.

**Comment by Jyc323** [comment]
Hi @sovrasov, thanks for your suggestion, I carefully thought about it. I refactor `OTXHlabelClsModel` by a dynamic wrapping approach to avoid modifying all child implementations or external calls.
In `base.py`, I added a custom `__getattribute__` hook that automatically wraps each child’s `_create_model()` at runtime. The wrapper injects a common post-processing step that executes immediately after `_create_model()` finishes.
The parent class now transparently runs an additional step, `_finalize_model()`, after model creation. `_finalize_model()` replaces the classifier with `KLHLabelClassifier` when `kl_weight > 0`, enabling hybrid training that combines the Tree-Path KL loss with standard cross-entropy.
Since `kl_weight` is a new hyperparameter, I added `**kwargs` parameter to `OTXHlabelClsModel`. This allows the class to accept additional optional arguments for future extensibility. However, I’m not sure if it reduces type safety or clarity.
I also add more unit tests based on the suggestions.
For the changed files, please take a look and let me know if it aligns with your expectations. If it does, I’ll proceed with adding  the related documentation

File: library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py [hunk 9]
```diff
@@ -89,3 +91,66 @@ def _create_model(self, head_config: dict | None = None) -> nn.Module:  # type:
...
```
Comment (sovrasov) [inline library/src/otx/backend/native/models/classification/hlabel_models/timm_model.py]: Should a specific classifier for timm be deleted since we have a common one?

File: library/src/otx/backend/native/models/classification/hlabel_models/base.py [hunk 10]
```diff
@@ -59,7 +61,9 @@ def __init__(
         scheduler: LRSchedulerCallable | LRSchedulerListCallable = DefaultSchedulerCallable,
         metric: MetricCallable = HLabelClsMetricCallable,
         torch_compile: bool = False,
+        **kwargs,
     ) -> None:
+        self.kl_weight = kwargs.get("kl_weight", 0.0)
```
Comment (sovrasov) [inline library/src/otx/backend/native/models/classification/hlabel_models/base.py]: That extra hidden argument looks too implicit: for instance `kl_weight` arg is hidden from the signature, therefore no doc and IDE can not give relevant hints.

Since we have `kl_weight` anyway, I propose to make it a normal argument and describe the related logic in the docstring

**Comment by sovrasov** [comment]
> Hi @sovrasov, thanks for your suggestion, I carefully thought about it. I refactor `OTXHlabelClsModel` by a dynamic wrapping approach to avoid modifying all child implementations or external calls. In `base.py`, I added a custom `__getattribute__` hook that automatically wraps each child’s `_create_model()` at runtime. The wrapper injects a common post-processing step that executes immediately after `_create_model()` finishes. The parent class now transparently runs an additional step, `_finalize_model()`, after model creation. `_finalize_model()` replaces the classifier with `KLHLabelClassifier` when `kl_weight > 0`, enabling hybrid training that combines the Tree-Path KL loss with standard cross-entropy. Since `kl_weight` is a new hyperparameter, I added `**kwargs` parameter to `OTXHlabelClsModel`. This allows the class to accept additional optional arguments for future extensibility. However, I’m not sure if it reduces type safety or clarity. I also add more unit tests based on the suggestions. For the changed files, please take a look and let me know if it aligns with your expectations. If it does, I’ll proceed with adding the related documentation

Thanks @Jyc323, I've left a couple of comments to the latest changes

**Comment by Jyc323** [comment]
Hi @sovrasov , I resolved the comment, and other errors from tox. Please have a review, thanks!

**Comment by sovrasov** [comment]
> Hi @sovrasov , I resolved the comment, and other errors from tox. Please have a review, thanks!

@Jyc323 it looks like ruff check is still failing:
```
ruff.....................................................................Failed
- hook id: ruff
- files were modified by this hook
```

**Comment by Jyc323** [comment]
Hi, @sovrasov, I have fixed the errors, please have a review, thanks

**Comment by sovrasov** [comment]
@Jyc323 now unit tests are failing

**Comment by Jyc323** [comment]
Hi @sovrasov, thanks for the review! The unit tests were passing on my side. Could you please recheck when you get a chance?

File: library/src/otx/backend/native/utils/utils.py [hunk 11]
```diff
@@ -86,7 +86,7 @@ def mock_modules_for_chkpt() -> Iterator[None]:
         sys.modules["otx.core.types.task"] = otx.types.task
         sys.modules["otx.core.types.label"] = otx.types.label
         sys.modules["otx.core.model"] = otx.backend.native.models  # type: ignore[attr-defined]
-        sys.modules["otx.core.metrics"] = otx.metrics
+        # sys.modules["otx.core.metrics"] = otx.metrics
```
Comment (leoll2) [inline library/src/otx/backend/native/utils/utils.py]: This change is probably causing this test to fail:

image

**Comment by Jyc323** [comment]
Hi @leoll2, thanks for your review. I restored the utils.py file now. Please let me know if there is any other issues.

**Comment by sovrasov** [comment]
Thanks a lot @Jyc323! From the code perspective, we are there, so now could you update the advanced h-cls tutorial you have added in the previous PR to avoid these improvements being silently buried inside the library?

**Comment by Jyc323** [comment]
Hi @sovrasov , docs have been updated: added the Tree-Path KL tutorial. Please review when convenient!

**Review by sovrasov — APPROVED** [review]

**Comment by sovrasov** [comment]
@Jyc323 thanks for the contribution, impressive work!
